crates/ty_python_semantic/resources/mdtest/type_properties/satisfied_by_all_typevars.md

```diff
@@ -0,0 +1,158 @@
+    static_assert(ConstraintSet.range(Never, T, Base).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Sub).satisfied_by_all_typevars(T))
+```
+
+If a typevar has an upper bound, then it must specialize to a type that is a subtype of that bound.
+For an inferable typevar, that means we need a single type that satisfies both the constraint set
+and the upper bound.
+
+```py
+def bounded[T: Base]():
+    static_assert(ConstraintSet.always().satisfied_by_all_typevars(T))
+    static_assert(not ConstraintSet.never().satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Super).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Base).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Sub).satisfied_by_all_typevars(T))
+
+    # This succeeds because T can specialize to Never
+    constraints = ConstraintSet.range(Never, T, Unrelated)
+    static_assert(constraints.satisfied_by_all_typevars(T))
+
+    # If we explicitly disallow Never, then it fails
+    constraints = constraints & ~ConstraintSet.range(Never, T, Never)
+    static_assert(not constraints.satisfied_by_all_typevars(T))
+```
+
+If a typevar has constraints, then it must specialize to one of those specific types. (Not to a
+subtype of one of those types!) For an inferable typevar, that means we need the constraint set to
+be satisfied by any one of the constraints.
+
+```py
+def constrained[T: (Base, Unrelated)]():
+    static_assert(ConstraintSet.always().satisfied_by_all_typevars(T))
+    static_assert(not ConstraintSet.never().satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Unrelated).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Super).satisfied_by_all_typevars(T))
+    static_assert(ConstraintSet.range(Never, T, Base).satisfied_by_all_typevars(T))
+    # Sub is not Base! Constraints must be exact, not subtypes
+    static_assert(not ConstraintSet.range(Never, T, Sub).satisfied_by_all_typevars(T))
+```
+
+## Non-inferable typevars
+
+When a typevar is in a non-inferable position, the constraint set must be satisfied for _every_
+valid specialization. The most common non-inferable position occurs in the body of a generic
+function or class: here we don't know in advance what type the typevar will be specialized to, and
+so we have to ensure that the body is valid for all possible specializations.
+
+```py
+from typing import final, Never
+from ty_extensions import ConstraintSet, static_assert
+
+class Super: ...
+class Base(Super): ...
+class Sub(Base): ...
+
+@final
+class Unrelated: ...
+```
+
+If a typevar has no bound or constraints, then it can specialize to any type. For a non-inferable
+typevar, that means the constraint set must be satisfied for every possible type.
+
+```py
+def unbounded[T]():
+    static_assert(ConstraintSet.always().satisfied_by_all_typevars())
```

Comment:
Gosh... it took me way too long to notice the difference to the unbounded example on line 40... We don't pass any type var here...

Should we use a different method name for this? It seems very subtle, but maybe it's obvious for someone more familiar with all this

Reply:
I can make it a keyword-only parameter, which I think will call out the difference more clearly